src/dune_pkg/rev_store.ml

```diff
@@ -391,22 +391,88 @@ module Entry = struct
   ;;
 end
 
-module Submodule = struct
-  (* a submodule in [.gitmodules] can also have a [branch] but given we only
-     need to resolve the commit object, we don't have to care about the
-     tracking branch *)
+module At_rev = struct
+  type repo = t
+
   type t =
-    { path : Path.Local.t
+    { repo : repo
+    ; revision : Rev.t
     ; source : string
+    ; files : File.Set.t
     }
 
-  let parse lines =
-    match Git_config_parser.parse lines with
-    | Error err ->
-      (* CR-rgrinberg: the loc needs to be pulled from the git URL *)
-      User_error.raise [ Pp.textf "Failed to parse submodules: %s" err ]
-    | Ok cfg ->
-      List.filter_map cfg ~f:(fun { name; arg = _; bindings } ->
+  module Config = struct
+    type bindings = string * string
+
+    type section =
+      { name : string
+      ; arg : string option
+      ; bindings : bindings list
+      }
+
+    type t = section list
+
+    module KV = struct
+      module T = struct
+        type t = string * string option
+
+        let compare = Tuple.T2.compare String.compare (Option.compare String.compare)
+        let to_dyn = Tuple.T2.to_dyn Dyn.string (Dyn.option Dyn.string)
+      end
+
+      include Comparable.Make (T)
+    end
+
+    let config { dir } revision path : t Fiber.t =
+      let (Rev.Rev rev) = revision in
+      let location = [ "--blob"; sprintf "%s:%s" rev (Path.Local.to_string path) ] in
+      let command = [ "config"; "--list" ] @ location in
+      let failure_mode =
+        Process.Failure_mode.Accept
+          (Predicate.create (fun x -> Int.equal x 0 || Int.equal x 128))
```

Comment:
What does 128 map to?

Reply:
This is what `git` returns when the blob is not found (e.g. because the revisions just doesn't have a `.gitmodules`).

Reply:
it would be useful to test for that situation separately, to avoid handling everything at once. do we have a working tree at that point? if that's the case, we can check for existence (and use `--file`). Otherwise, we can use `git cat-file -t main:.gitsubmodules`.

Or actually, maybe just use a failure mode of `Return` and deal with it at that point (`Error 128 -> []`).

Reply:
Good point, it's a bit of a shame that in any case the code is 128. I've split it up to check for the existence of the file with `cat-file -e` and used `run_capture_lines` which displays errors on failures.